File: LC0152_Max_Product_Subarray/max_product_subarray.go
package main

import (
	"math"
)

func maxProduct(nums []int) int {
	n := len(nums)
	// trivial case
	if n == 1 {
		return nums[0]
	}

	maxProd, minProd, result := nums[0], nums[0], nums[0]

	for i := 1; i < n; i++ {
		if nums[i] < 0 {
			maxProd, minProd = minProd, maxProd
		}
		maxProd = int(math.Max(float64(nums[i]), float64(maxProd*nums[i])))
		minProd = int(math.Min(float64(nums[i]), float64(minProd*nums[i])))
		result = int(math.Max(float64(result), float64(maxProd)))
	}

	return result
}

File: LC0152_Max_Product_Subarray/max_product_subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]

        max_prod, min_prod, result = nums[0], nums[0], nums[0]
        for i in range(1, n):
            if nums[i] < 0:
                max_prod, min_prod = min_prod, max_prod
            max_prod = max(nums[i], max_prod * nums[i])
            min_prod = min(nums[i], min_prod * nums[i])
            result = max(result, max_prod)
        return result
        